Add support for mutual recursion

parno opened the pull request:
This PR adds support for mutually recursive function calls, and incidentally allows function calls to functions defined later in the file (i.e., we no longer require the functions in the file to be topologically sorted).

parno committed:
Create a basic test case for mutual recursion

parno committed:
Initial version of code to compute SCCs for mutual recursion detection.

parno committed:
Initial progress towards creating a function call graph.

parno committed:
Change SCC to require Clone instead of Copy.  That plus lots
of calls to clone() allows the initial construction of the call graph to
compile.

parno committed:
Use the call graph's SCCs to check for mutual recursion

parno committed:
Expand our basic test and check for mutual recursion in statements too.
Still need to use mutual recursion info when creating AIR.

parno committed:
Small clean up

parno committed:
Pre-declare spec function symbols, so that function definitions need
not occur in order in the file.

parno committed:
Pre-declare recursive versions too, so other members of the SCC (if any)
can refer to them.

parno committed:
When emitting the body for an element of an SCC, use the recursive name
for all calls to other members of the SCC

parno committed:
Expand unit test

parno committed:
Add a test case illustrating the need for better decreases checking of
mutually recursive functions

parno committed:
Decreases check needs to check for calls into same SCC, not just the
function itself.  Result is still problematic, since our tricky test
case still passes.

parno committed:
Terminates now uses the decreases clause from the callee,
which required changing the return type of terminates to admit the
possibility of error

parno committed:
Clean up suggested by rust-fmt and clippy

parno commented:
I'm still pretty new to Rust, so let me know if I've done something that's not Rust-like.  Jay pointed me at the [clippy](https://github.com/rust-lang/rust-clippy) tool as a guide to writing idiomatic Rust.  Not sure if we want to use that in addition to rust-fmt, but it seems worth considering.

parno commented:
Also, what's our policy on unit tests?  At the moment, we have files in `verify/rust_verify/example/*rs`, which is where I've put my new unit tests for mutual recursion (https://github.com/Chris-Hawblitzel/rust/pull/19/files#diff-70c5ca6781df7cf0a113118ecd12c85782abc9485ab60c3fc3dfb632cc4badf7).   But we also have `verify/rust_verify/tests/*rs`.  Any thoughts on the division of labor between the two?  The latter seems like a better home moving forward, especially since it supports negative tests cleanly.  I currently have a negative test in `verify/rust_verify/example/recursion.rs`, but that may be a bad home for it.

parno commented:
Finally, I couldn't find a good way to test mutually recursive statements.  If I define:

```
#[spec]
fn count_down_a_stmt_tricky(i:nat) -> nat {
    decreases(i);

    if i == 0 {
        0
    } else {
        let r = count_down_b_stmt_tricky(i + 1);
        4
    }
}

#[spec]
fn count_down_b_stmt_tricky(i:nat) -> nat {
    decreases(i);

    if i == 0 { 
        0
    } else {
        let r = count_down_a_stmt_tricky(i + 1);
        6
    }
}
```
Then the calls are treated as expressions.  If I remove the `let r =`, then Dust panics with a `not yet implemented` error.  If I change the function types to not return anything, then Dust ignores the two functions (i.e., doesn't emit anything to AIR), because the function-related AIR translation steps, e.g., `func_decl_to_air` all seem to be predicated on having `Some` return value, e.g., 
```
match (function.x.mode, function.x.ret.as_ref()) {
        (Mode::Spec, Some((_, ret, _))) => {
```
Is that a concern?  The same thing happens even for self-recursion; i.e., Dust doesn't object to:
```
#[spec]
fn count_down_tricky(i:nat) {
    decreases(i);

    if i != 0 {
        count_down_tricky(i + 1)
    }
}
```

Chris-Hawblitzel commented:
For testing statements, you can use a #[proof] function instead of a #[spec] function.  Right now, #[spec] functions can only have expression-like bodies (e.g. no while loops, no updates to local variables, etc.), and this will probably be the case for a while, since this makes it easy to translate #[spec] functions into Z3 functions.

Chris-Hawblitzel commented:
The `#[spec] fn count_down_tricky(i:nat) {` function is interesting.  Right now, we treat Rust's "DefaultReturn" as meaning "return nothing".  But really, it means return Rust's unit value (`fn f() -> () {...)` and `fn f() {...}` are equivalent in Rust).  #[spec] functions are only useful for their return value, so it's not clear anyone needs a function that returns the unit type, since any such function can only return exactly the unit value, but I don't see any reason to rule it out either.  Let me make this a work item for myself: I'll take a look at refactoring VIR so that we use the unit return type instead of no return type for functions returning unit/nothing.

Chris-Hawblitzel commented:
Now that we have Andrea's test framework, it would be great if we could write lots of positive unit tests and even more negative unit tests in the `tests` directory.  Going forward, we can use the `example` directory as more of a feature tutorial for people who are new to the language and want to see examples of various features.

parno committed:
Expand tests to include mutually recursive statements, and update the
implementation to account for them.

parno committed:
Formatting

parno commented:
Ah, good point on testing statements with proof functions.  I added some tests of that sort and updated the implementation to handle them properly.

Thanks for taking a look at the unit return value item.

I'll work on splitting the current `examples/recursion.rs` into a few didactic examples, and move the rest of the test cases, plus some others I've been thinking about, over into `tests`.

parno committed:
Refactor most of the unit tests into the tests directory

parno committed:
Make things more succinct and Rust-like

parno committed:
Add some additional tests

utaal commented on verify/vir/src/recursion.rs:
```diff
@@ -58,63 +62,76 @@ fn check_decrease_rename(ctxt: &Ctxt, span: &Span, args: &Exps) -> Exp {
 }
 
+fn update_decreases_exp<'a>(ctxt: &'a Ctxt, name: &Ident) -> Result<Ctxt<'a>, VirErr> {
+    let function = ctxt.ctx.func_map.get(name).unwrap();
+    let (new_decreases_expr, _) = function.x.decrease.as_ref().unwrap().clone();
```
I tend to be careful with `unwrap`s, except in cases where it's obvious that it's not going to fail (e.g. I checked `.is_some()` a few lines above). If that's the case here, it may be good to add a comment; if not, I'd recommend using `expect` instead, and providing a message that describes what unexpected condition caused the error. (Here, I imagine, we expected for the function to be in the `func_map`, but it wasn't.)

parno replied:
I'm not sure what would be a useful error message in these cases.  Both indicate an internal error that the user can't address.    On line 66, `func_map` is supposed to hold all of the functions.  If it doesn't, we're in trouble.  On the next line, the function is `update_decreases_exp`, so it shouldn't be used on functions that don't have decreases clauses.  I could add  "function should have had a decreases clause", but I'm not sure if that add clarity.  If you think something along those lines would help or if you had something else in mind, we can certainly add it.

parno replied:
Okay, added some explicit text here.

utaal commented on verify/vir/src/scc.rs:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+
+    pub fn get_scc_size(&self, t: &T) -> usize {
```
Should there be an `assert(self.has_run)` here, and in `get_scc_rep`?

parno replied:
Yes, good point.

parno replied:
Fixed.

utaal commented:
>  it would be great if we could write lots of positive unit tests and even more negative unit tests in the tests directory. Going forward, we can use the `example` directory as more of a feature tutorial for people who are new to the language and want to see examples of various features.

I like this approach, and it's in line with what folks generally expect from rust codebases.

utaal commented on verify/vir/src/func_to_air.rs:
```diff
@@ -99,7 +99,7 @@ fn func_body_to_air(
         let mut args_fuel = args.clone();
         let mut args_succ = args.clone();
-        let mut args_def = args.clone();
+        let mut args_def = args;
```
Regarding `clippy`, I don't have strong opinions, except for a general bias towards minimising bureaucracy: I think folks learn how to write idiomatic rust by looking at idiomatic rust more than by being guided by tools. This change here for example I think is unnecessary, dropping `args` on the floor isn't going to meaningfully affect performance, and it's just going to make it minorly inconvenient for the next person that adds an `args` clone to this.
That said I'm not against trying out `clippy`, just not particularly excited :)

parno replied:
Fair enough.  Let's chat with Chris when he returns and see if he has a preference.  I don't feel strongly either way.

utaal commented on verify/vir/src/scc.rs:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+
+    pub fn compute_sccs(&mut self) {
```
For things like this, when convenient, I try to encode the state machine states in different types. Here this could be:
```
pub fn compute_sccs(self) -> ConnectedComponents {
...
```
that returns a new struct with the compted sccs and any other info it needs to answer the queries (`get_scc_rep`, `get_scc_size`). Note that the proposed signature consumes self. Unless I'm missing something, this would let you get rid of the runtime assertions (around `has_run`).

parno replied:
Interesting.  I think what makes that tricky here is that the `Graph` is an element of the global `Ctx` that's passed around everywhere.  It's in there so that we can build up the call graph in the first place.  Similarly, we want to be able to query for SCC info all over the place.  If we split them, I think we would end up with two fields in `Ctx`, one that starts off with a valid Graph and gives it up, and one that starts off uninitialized and then gets filled in with Ctx.  Is there a more elegant way to handle that?

parno replied:
I guess another option would be to have the query methods each check whether SCC computation has occurred, and if not, run it.  That would remove the need for the asserts and probably wouldn't be any more expensive.

parno replied:
I started to implement the last option, but then I realized that `get_scc_size` and `get_scc_rep` would need to take `&mut self` instead of `&self`, which seems wrong for methods that are morally read-only, so I think I'm going to leave this as is for now.

parno committed:
Make sure SCC computation has run before answering queries.

parno committed:
Provide messages via expect, instead of unwrap
